File: Core/src/test/java/com/github/ezauton/core/simulator/SimulatedTankBotTest.java
package com.github.ezauton.core.simulator;

import com.github.ezauton.core.action.*;
import com.github.ezauton.core.actuators.VelocityMotor;
import com.github.ezauton.core.localization.estimators.TankRobotEncoderEncoderEstimator;
import com.github.ezauton.core.pathplanning.PP_PathGenerator;
import com.github.ezauton.core.pathplanning.Path;
import com.github.ezauton.core.pathplanning.purepursuit.Lookahead;
import com.github.ezauton.core.pathplanning.purepursuit.LookaheadBounds;
import com.github.ezauton.core.pathplanning.purepursuit.PPWaypoint;
import com.github.ezauton.core.pathplanning.purepursuit.PurePursuitMovementStrategy;
import com.github.ezauton.core.robot.implemented.TankRobotTransLocDriveable;
import com.github.ezauton.core.simulation.SimulatedTankRobot;
import com.github.ezauton.core.simulation.TimeWarpedSimulation;
import com.github.ezauton.core.trajectory.geometry.ImmutableVector;
import com.github.ezauton.core.utils.TimeWarpedClock;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class SimulatedTankBotTest {
    @Test
    public void testStraight2() throws IOException, TimeoutException, ExecutionException {
        PPWaypoint waypoint1 = PPWaypoint.simple2D(0, 0, 0, 3, -4);
        PPWaypoint waypoint2 = PPWaypoint.simple2D(0, 6, 1, 3, -4);
        PPWaypoint waypoint3 = PPWaypoint.simple2D(0, 20, 0, 3, -4);

        PP_PathGenerator pathGenerator = new PP_PathGenerator(waypoint1, waypoint2, waypoint3);

        Path path = pathGenerator.generate(0.05);

        PurePursuitMovementStrategy ppMoveStrat = new PurePursuitMovementStrategy(path, 8);

        TimeWarpedClock clock = new TimeWarpedClock(10);
        SimulatedTankRobot bot = new SimulatedTankRobot(0.2, clock, 3, 0.2, 4);
        bot.getDefaultLocEstimator().reset();
        VelocityMotor leftMotor = bot.getLeftMotor();
        VelocityMotor rightMotor = bot.getRightMotor();

        TankRobotEncoderEncoderEstimator locEstimator = new TankRobotEncoderEncoderEstimator(bot.getLeftDistanceSensor(), bot.getRightDistanceSensor(), bot);
        locEstimator.reset();

        TimeWarpedSimulation sim = new TimeWarpedSimulation(10);

        BackgroundAction background = new BackgroundAction(50, TimeUnit.MILLISECONDS, bot::update, locEstimator::update);

        Lookahead lookahead = new LookaheadBounds(1, 5, 2, 10);

        TankRobotTransLocDriveable tankRobotTransLocDriveable = new TankRobotTransLocDriveable(leftMotor, rightMotor, locEstimator, locEstimator, bot);

        PurePursuitAction purePursuitAction = new PurePursuitAction(50, TimeUnit.MILLISECONDS, ppMoveStrat, locEstimator, locEstimator, lookahead, tankRobotTransLocDriveable);

        ActionGroup actionGroup = new ActionGroup()
                .with(background)
                .addSequential(purePursuitAction);

        sim.add(actionGroup);


        sim.runSimulation(12, TimeUnit.SECONDS);

    }

    @Test
    public void testStraight() throws TimeoutException, ExecutionException {

        TimeWarpedSimulation sim = new TimeWarpedSimulation();
        SimulatedTankRobot simulatedBot = new SimulatedTankRobot(0.2, sim.getClock(), 3, -4, 4);
        simulatedBot.getDefaultLocEstimator().reset();
        TankRobotEncoderEncoderEstimator locEstimator = new TankRobotEncoderEncoderEstimator(simulatedBot.getLeftDistanceSensor(), simulatedBot.getRightDistanceSensor(), simulatedBot);
        locEstimator.reset();

//        sim.add(new TimedPeriodicAction(5, TimeUnit.SECONDS, () -> simulatedBot.run(1, 1)));
//
//        sim.add(new BackgroundAction(10, TimeUnit.MILLISECONDS, locEstimator::update, simulatedBot::update));

        ActionGroup actionGroup = new ActionGroup()
                .addParallel(new TimedPeriodicAction(5, TimeUnit.SECONDS, () -> simulatedBot.run(1, 1)))
                .with(new BackgroundAction(10, TimeUnit.MILLISECONDS, locEstimator::update, simulatedBot::update))
                .addSequential(new DelayedAction(7, TimeUnit.SECONDS));

        sim.add(actionGroup);

        sim.runSimulation(10, TimeUnit.SECONDS);

        simulatedBot.run(0, 0);

        final ImmutableVector estimatedLocation = locEstimator.estimateLocation();

        assertTrue(estimatedLocation.get(1) > 5);
        assertTrue(Math.abs(estimatedLocation.get(0)) < 0.1);

    }
}

File: Core/src/test/kotlin/com/github/ezauton/core/simulator/SimulatedTankBotTest.kt
package com.github.ezauton.core.simulator

import com.github.ezauton.conversion.*
import com.github.ezauton.core.action.*
import com.github.ezauton.core.localization.estimators.TankRobotEncoderEncoderEstimator
import com.github.ezauton.core.pathplanning.TrajectoryGenerator
import com.github.ezauton.core.pathplanning.purepursuit.LookaheadBounds
import com.github.ezauton.core.pathplanning.purepursuit.PPWaypoint
import com.github.ezauton.core.robot.implemented.TankRobotTransLocDrivable
import com.github.ezauton.core.simulation.SimulatedTankRobot
import com.github.ezauton.core.simulation.parallel
import com.github.ezauton.core.simulation.sequential
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.io.IOException
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeoutException

class SimulatedTankBotTest {
  @Test
  @Throws(IOException::class, TimeoutException::class, ExecutionException::class)
  fun testStraight2() = runBlocking{

    val waypoint1 = PPWaypoint.simple2D(0.m, 0.m,0.mps, 3.mpss, (-4.0).mpss)
    val waypoint2 = PPWaypoint.simple2D(0.m, 6.m,1.mps, 3.mpss, (-4.0).mpss)
    val waypoint3 = PPWaypoint.simple2D(0.m, 20.m, 0.mps, 3.mpss, (-4.0).mpss)

    val pathGenerator = TrajectoryGenerator(waypoint1, waypoint2, waypoint3)

    val trajectory = pathGenerator.generate(0.05.seconds)

    val bot = SimulatedTankRobot.create(0.2.m, 3.0.mpss, 0.2.mps, 4.0.mps)

    val leftMotor = bot.leftMotor
    val rightMotor = bot.rightMotor

    val locEstimator = TankRobotEncoderEncoderEstimator.from(bot.leftDistanceSensor, bot.rightDistanceSensor, bot)
//    locEstimator.reset()


    val background = periodicAction(50.ms) {
      bot.update()
      locEstimator.update()
      return@periodicAction
    }

    val lookahead = LookaheadBounds(1.0.m, 5.0.m, 2.0.mps, 10.0.mps, locEstimator)

    val drivable = TankRobotTransLocDrivable(leftMotor, rightMotor, locEstimator, locEstimator, bot)

    val purePursuitAction = purePursuit(Period(50.ms), trajectory, locEstimator, drivable , lookahead, stopDistance = 8.m)

    val actionGroup = action {
      ephemeralScope {
        parallel(background)
        sequential(purePursuitAction)
      }
    }

    actionGroup.runWithTimeout(10.seconds)
  }

  @Test
  @Throws(TimeoutException::class, ExecutionException::class)
  fun testStraight() = runBlocking {

    val simulatedBot = SimulatedTankRobot.create(0.2.m, 3.0.mpss, (-4.0).mps, 4.0.mps)

    val actionGroup = action {
      parallel {
        periodic(duration = 5.seconds){
          simulatedBot.run(1.0.mps, 1.0.mps)
        }
      }
    }


    actionGroup.runWithTimeout(10.seconds)

    // stop the robot
    simulatedBot.run(0.0.mps, 0.0.mps)

    val estimatedLocation = simulatedBot.estimateLocation()

    println("estimated loc: $estimatedLocation")
    assertTrue(estimatedLocation.y > 4.5.m)
    assertTrue(estimatedLocation.y < 5.5.m)
    assertTrue(estimatedLocation.x.abs() < 0.1.m)

  }
}
